release-21.1: changefeedccl: Add a mechanism to throttle buffer drain rate.

miretskiy opened the pull request:
Backport 1/1 commits from #67169.

/cc @cockroachdb/release

---

Make it possible to specify per-node changefeed configuration
`changefeed.node_sink_throttle_config` which enables throttling
the rate of drain of the memory buffer.

The throttling can be configure to limit the rate of message
emmission, the size of the emitted messages, as well as the
sink flush counts.  All of the throttles are specified as a
base value in units per second as well as burst rate:

```
{"MessageRate": 100, "MessageBurst": 200, "FlushRate": 1},
```

If the value is unspecified, the throttling is disabled.
The throttling configuration does not imply exact match on the
observed emitted values.  This is primarily due to per-sink
buffering behavior.  However, the throttling apply to the
buffer drain rate will have proportinoal impact on the
traffic emitted to the downstream sink.

When throttling is active, and the traffic is throttled, the changefeed
enters a "pushback" mode.  If the throttling configuration is such
that the traffic throttled below the incoming rate of messages,
the changefeed will deal with this situation by eventually restarting
the rangefeeds and performing a catchup scan.  Note, however, that such
regime is detrimental for overall performance of changefeeds.
Throttling should be used in order to limit short burst of traffic.

Follow on PRs will add throttling to the sinks themselves, as well
as add monitoring metrics.

Informs cockroachdb#58967

Release Notes: Changefeeds can be configured to throttle the rate
of emission from the memory buffer thus making it possible to limit
the emission rate from the changefeeds.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68628)
<!-- Reviewable:end -->

miretskiy commented:
FYI: This was not a clean backport; had to backout of renames kvfeed.Event -> kvevent.Event; plus others.
Please review more thoroughly than regular backports.

Commit:
changefeedccl: Add a mechanism to throttle buffer drain rate.

Make it possible to specify per-node changefeed configuration
changefeed.node_sink_throttle_config which enables throttling
the rate of drain of the memory buffer.

The throttling can be configure to limit the rate of message
emmission, the size of the emitted messages, as well as the
sink flush counts. All of the throttles are specified as a
base value in units per second as well as burst rate:

{"MessageRate": 100, "MessageBurst": 200, "FlushRate": 1},
If the value is unspecified, the throttling is disabled.
The throttling configuration does not imply exact match on the
observed emitted values. This is primarily due to per-sink
buffering behavior. However, the throttling apply to the
buffer drain rate will have proportinoal impact on the
traffic emitted to the downstream sink.

When throttling is active, and the traffic is throttled, the changefeed
enters a "pushback" mode. If the throttling configuration is such
that the traffic throttled below the incoming rate of messages,
the changefeed will deal with this situation by eventually restarting
the rangefeeds and performing a catchup scan. Note, however, that such
regime is detrimental for overall performance of changefeeds.
Throttling should be used in order to limit short burst of traffic.

Follow on PRs will add throttling to the sinks themselves, as well
as add monitoring metrics.

Informs #58967

Release Notes: Changefeeds can be configured to throttle the rate
of emission from the memory buffer thus making it possible to limit
the emission rate from the changefeeds.

HonoreDB reviewed (approved):
LGTM!

> Reviewed 12 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68628#-:-MglB1T27KcbyIeRMvZX:b-85i648)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)



<!-- Sent from Reviewable.io -->
